Updates to the ESC environments docs (#3595)
pulumi/pulumi-hugo: themes/default/content/docs/pulumi-cloud/esc/environments.md

~
~
~
"/tmp/2198639483.yaml" 29L, 528B
```

> You can also pass an explicit editor using the `--editor` flag:

Author: This tells VS Code to stay open while the environment is edited. I'm doublechecking this is still necessary (it used to be, but at least for me today, no longer seems to be); will remove if it isn't.

> You can also pass an explicit editor using the `--editor` flag:

Author: [Asked here](https://pulumi.slack.com/archives/C05G9TV06MB/p1699384341915099), let's see how it shakes out. In the meantime, I'll leave it, in case.

> You can also pass an explicit editor using the `--editor` flag:

Author: Scott tried also and confirmed it's no longer needed, so I'll remove. I am curious as to what exactly changed, though! 🤔

> You can also pass an explicit editor using the `--editor` flag:

Author: Pat just confirmed:

> if the editor is exactly code we automatically pass -w

And indeed: https://github.com/pulumi/esc/blob/38cfd9d798c030af3ab66b9e7166cf60296564e4/cmd/esc/cli/env_edit.go#L177-L180

        oidc:
          roleArn: arn:aws:iam::01234567891011:role/some-role
          sessionName: some-session

  environmentVariables:
    AWS_ACCESS_KEY_ID: ${aws.login.accessKeyId}
    AWS_SECRET_ACCESS_KEY: ${aws.login.secretAccessKey}
    AWS_SESSION_TOKEN: ${aws.login.sessionToken}
    MY_ENV_VAR: "true"
```

To render these values as environment variables in the shell, use `esc env open`, passing either `--format shell` or `--format dotenv`:

```bash
$ esc env open myorg/test --format shell
export AWS_ACCESS_KEY_ID="AKIAIOSFODNN7EXAMPLE"
export AWS_SECRET_ACCESS_KEY="wJalrXUtnFEMI/K7MDENG/bPxRfiCYEXAMPLE"
export AWS_SESSION_TOKEN="eyJpc3MiOiJPbmxpbmUgSldUIEJ1aWxkZXIiLCJpYXQiOEXAMPLE"
export MY_ENV_VAR="true"

$ esc env open myorg/test --format dotenv
AWS_ACCESS_KEY_ID="AKIAIOSFODNN7EXAMPLE"
AWS_SECRET_ACCESS_KEY="wJalrXUtnFEMI/K7MDENG/bPxRfiCYEXAMPLE"
AWS_SESSION_TOKEN="eyJpc3MiOiJPbmxpbmUgSldUIEJ1aWxkZXIiLCJpYXQiOEXAMPLE"
MY_ENV_VAR="true"
```

To export rendered variables into the current shell, use `eval` or similar:

```yaml
$ eval $(esc env open myorg/test --format shell)
$ echo $MY_ENV_VAR
true
```

## Running commands with environment variables

> You can also run CLI commands directly, using environment variables obtained with Pulumi ESC --- without having to export those variables into the shell first.

Author: Hm. Well, considering this section is just about `esc run` and environment variables in general, I'm inclined to leave it as-is, since it does seem to work as described (both in this example and other commands). I've seen some discussion of oddness running `pulumi up` with Azure OIDC, though. Is that what you mean?

> You can also run CLI commands directly, using environment variables obtained with Pulumi ESC --- without having to export those variables into the shell first.

Author: Merging to get the rest of these updates out, but happy to address any follow-ups that come out this thread. (Thanks @toriancrane!)

To do this, use `esc run <environment-name> <command>`:

```bash
$ esc run myorg/test aws s3 ls
2023-10-10 16:09:19 my-s3-bucket
```

If you need to pass one or more flags to the command, prefix the command with `--`:

```bash
$ esc run myorg/test -- aws s3 ls s3://my-s3-bucket --recursive --summarize
...
Total Objects: 5087
   Total Size: 2419123156
```

For additional options and details, see `esc run --help`.

> ## Composing environments with imports

Author: Yep, works for me!

In the following example, two environments, `aws-dev` and `stripe-dev`, are used to compose a third environment, `myapp-dev`:

```yaml
# myorg/aws-dev
values:
  aws:
    region: us-west-2
    login:
      fn::open::aws-login:
        static:
          accessKeyId:
            fn::secret: AKIAIOSFODNN7EXAMPLE
          secretAccessKey:
            fn::secret: wJalrXUtnFEMI/K7MDENG/bPxRfiCYEXAMPLE
```

```yaml
# myorg/stripe-dev
values:
  stripe:
    apiURL: https://api.stripe.com
    apiKey:
      fn::secret: sk_XemWAl12i4x3hZhp4vBKDEXAMPLE
```

The application-specific `myapp-dev` environment then `imports` these two environments, inheriting their settings automatically:

```yaml
# myorg/myapp-dev
imports:
  - aws-dev
  - stripe-dev

values:
  greeting: Hello from the dev environment!

  environmentVariables:
    AWS_ACCESS_KEY_ID: ${aws.login.accessKeyId}
    AWS_SECRET_ACCESS_KEY: ${aws.login.secretAccessKey}
    STRIPE_API_KEY: ${stripe.apiKey}
    STRIPE_API_URL: ${stripe.apiURL}
    GREETING: ${greeting}
```

> Use `esc run` to compose the `myapp-dev` environment from the `aws-dev` and `stripe-dev` environments and expose its values to any application that needs it:

Author: Yep, works for me!

```bash
$ esc run myorg/myapp-dev -- bash -c 'echo $GREETING'
Hello from the dev environment!

$ esc run myorg/myapp-dev -- bash -c 'echo $STRIPE_API_URL'
https://api.stripe.com

$ esc run myorg/myapp-dev -- bash -c 'echo $STRIPE_API_KEY'
[secret]

$ esc run myorg/myapp-dev -- bash -c 'echo $AWS_SECRET_ACCESS_KEY'
[secret]

$ echo "'$GREETING'"
''
```

Notice in the example that the `environmentVariables` were exposed to the `bash` command, but not to the surrounding shell, and that the values marked as secrets with `fn::secret` were protected from exposure.

## Using environments with Pulumi IaC

With support for Pulumi ESC built into the Pulumi CLI, you can expose an environment's settings and secrets directly to any or all of your Pulumi stacks, bypassing the need to define and maintain individual configuration settings or secrets "locally" in Pulumi stack-configuration files. The optional `pulumiConfig` key enables this.

> The following example extends the `myorg/myapp-dev` environment, adding a `pulumiConfig` block to specify the settings to expose through standard [Pulumi configuration](/docs/concepts/config/):

Reviewer: ```suggestion
The following example updates the `myorg/myapp-dev` environment by adding a `pulumiConfig` block. This block specifies the [Pulumi config](/docs/concepts/config/) settings to expose to Pulumi IaC:
```
